### PR TITLE
[feat] refactor vilbert to make it compatible for torchscript

### PR DESCRIPTION
**Pull request by neo315**
Summary:
Refactor to make vilbert compatible to torchscript
Similar changes are done in [our implementation of Vilbert](https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/fbcode/assistant/mmu/multimodal_bert/vilbert/model.py). ref diff: D18892981

Differential Revision: D23923842

### CODE REVIEW

File: mmf/models/vilbert.py [hunk 1]
```diff
@@ -1,18 +1,21 @@
+#!/usr/bin/env python3
```
Comment (vedanuj) [inline mmf/models/vilbert.py:1]: Remove

File: mmf/models/vilbert.py [hunk 2]
```diff
@@ -53,10 +56,10 @@ def transpose_for_scores(self, x):
             self.num_attention_heads,
             self.attention_head_size,
         )
-        x = x.view(*new_x_shape)
+        x = torch.reshape(x, new_x_shape)
```
Comment (vedanuj) [inline mmf/models/vilbert.py:59]:
```suggestion
        x = x.view(new_x_shape)
```

Using view is better as reshaping might make a copy of the tensor in some cases.

File: mmf/models/vilbert.py [hunk 3]
```diff
@@ -83,7 +86,7 @@ def forward(self, hidden_states, attention_mask):
         context_layer = torch.matmul(attention_probs, value_layer)
         context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
         new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
-        context_layer = context_layer.view(*new_context_layer_shape)
+        context_layer = torch.reshape(context_layer, new_context_layer_shape)
```
Comment (vedanuj) [inline mmf/models/vilbert.py:89]:
```suggestion
        context_layer = context_layer.view(new_context_layer_shape)
```

File: mmf/models/vilbert.py [hunk 4]
```diff
@@ -324,18 +361,18 @@ def transpose_for_scores(self, x):
             self.num_attention_heads,
             self.attention_head_size,
         )
-        x = x.view(*new_x_shape)
+        x = torch.reshape(x, new_x_shape)
```
Comment (vedanuj) [inline mmf/models/vilbert.py:364]:
```suggestion
        x = x.view(new_x_shape)
```

File: mmf/models/vilbert.py [hunk 5]
```diff
@@ -377,7 +414,7 @@ def forward(
         context_layer1 = torch.matmul(attention_probs1, value_layer1)
         context_layer1 = context_layer1.permute(0, 2, 1, 3).contiguous()
         new_context_layer_shape1 = context_layer1.size()[:-2] + (self.all_head_size,)
-        context_layer1 = context_layer1.view(*new_context_layer_shape1)
+        context_layer1 = torch.reshape(context_layer1, new_context_layer_shape1)
```
Comment (vedanuj) [inline mmf/models/vilbert.py:417]: Replace `reshape` to `view`

File: mmf/models/vilbert.py [hunk 6]
```diff
@@ -835,19 +918,38 @@ def __init__(self, config):
...
+    @torch.jit.unused
+    def _change_dtype_for_fp16(self, extended_attention_mask):
+        """
+        change attention tensor's precision for fp16 training only
+        """
+        if self.training:
+            extended_attention_mask = extended_attention_mask.to(
+                dtype=next(self.parameters()).dtype
+            )  # fp16 compatibility
+        return extended_attention_mask
```
Comment (vedanuj) [inline mmf/models/vilbert.py:930]: This is not required with latest pytorch 1.6+. We can guard blocks using 'torch.jit.is_scripting()'. Check the implementation in VisualBERT model.

File: mmf/models/vilbert.py [hunk 7]
```diff
@@ -877,18 +984,18 @@ def forward(
         # positions we want to attend and -10000.0 for masked positions.
         # Since we are adding it to the raw scores before the softmax, this is
         # effectively the same as removing these entirely.
-        extended_attention_mask = extended_attention_mask.to(
-            dtype=next(self.parameters()).dtype
-        )  # fp16 compatibility
+        if self.training:
+            extended_attention_mask = self._change_dtype_for_fp16(
+                extended_attention_mask
+            )
```
Comment (vedanuj) [inline mmf/models/vilbert.py:990]: Use `not torch.jit.is_scripting()`

File: mmf/models/vilbert.py [hunk 8]
```diff
@@ -877,18 +984,18 @@ def forward(
...
         extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
-
-        extended_attention_mask2 = extended_attention_mask2.to(
-            dtype=next(self.parameters()).dtype
-        )  # fp16 compatibility
-
-        extended_image_attention_mask = extended_image_attention_mask.to(
-            dtype=next(self.parameters()).dtype
-        )  # fp16 compatibility
+        if self.training:
```
Comment (vedanuj) [inline mmf/models/vilbert.py:992]: Same

**Review by vedanuj — CHANGES_REQUESTED** [review]
Thanks for making these changes. Left some comments.

**Review by vedanuj — APPROVED** [review]

**Review by apsdehal — APPROVED** [review]

**Pull request closed without merging**